### PR TITLE
Redesign the application choices page - Add grouped tabs under "Your applications"

### PR DESCRIPTION
**Pull request by tomas-stefano**
## Context

As part of the new applications, and combination of status groupings, we need to incorporate a tab system, for candidates to easily sort through applications that they want to see.

This PR includes a tab system, similar to the one we have on the ITT report with relevant groupings.

## Changes proposed in this pull request

Screenshot 2024-02-16 at 12 14 24
Screenshot 2024-02-16 at 12 14 15

## Guidance to review

1. Different browser testing
2. See on mobile and multiple screen readers
3. Is it showing only the applications in each grouping tabs (e.g "Draft" only showing drafts apps, etc)
4. Click on an application that has an offer should direct you to offer details page
5. When you have multiple interviews, it should show all of them
6. When you have multiple interviews and it cancels the last one, the cancelled interview shouldn't be shown
7. When you enter an inactive application and you can add more choices you should see a warning text
8. When you enter an inactive application and you can't add more choices you shouldn't see a warning text

## Link to Trello card

Because I started the tab on this PR some of edge cases and scenarios appeared. 

I decided to continue to do this on the same PR but I create multiple cards for the bugs/problems we encountered:

https://trello.com/c/9dXrWmAA/1180-apply-application-page-updated-tabs-and-components-part-63
https://trello.com/c/prYYDgh3/1313-apply-interviewing-application-page-fixes
https://trello.com/c/PI6TtVAI/1314-apply-show-offer-application-page-when-candidate-clicks-the-application-link-in-application-choices-page
https://trello.com/c/hW8cVyei/1307-apply-inactive-application-page-message-addition

### CODE REVIEW

File: app/frontend/styles/components/_tabs.scss [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      text-decoration: none;
+
+      &:focus {
+        padding-bottom: govuk-spacing(2);
+      }
+
+    }
+  }
+
+  .tabs-component-navigation__link[aria-current]::before {
```
Comment (tomas-stefano) [inline app/frontend/styles/components/_tabs.scss:54]: I had to do this was due to stylelint issues. More details see here: 085ea38fee9104702fa5146ba79e372d054ff0d4
If you have another way of doing this without the stylelint issues, I am happy to fix!

File: app/services/candidate_interface/sort_application_choices.rb [hunk 2]
```diff
@@ -4,19 +4,28 @@ def self.call(application_choices:)
...
           SELECT application.*,
+          CASE
+            WHEN (status IN ('offer')) THEN 'offers_received'
+            WHEN (status IN ('unsubmitted', 'application_not_sent', 'cancelled')) THEN 'draft'
+            WHEN (status IN ('rejected', 'conditions_not_met')) THEN 'unsuccessful'
+            WHEN (status IN ('interviewing', 'inactive', 'awaiting_provider_decision')) THEN 'in_progress'
+            WHEN (status IN ('offer_withdrawn', 'withdrawn')) THEN 'withdraw'
+            WHEN (status IN ('declined')) THEN 'declined'
+            ELSE ''
+          END AS application_choices_group,
```
Comment (tomas-stefano) [inline app/services/candidate_interface/sort_application_choices.rb:15]: I chosen consciously to add this to not rely on application_choices_group_number below. With application_choices_group_number made the code more difficult to read so I consciously add the application_choices_group (see commit history for more details). Putting the groups here made the code very easy to understand up to the `ApplicationChoiceListComponent` in my opinion.

File: app/frontend/styles/components/_tabs.scss [hunk 3]
```diff
@@ -0,0 +1,64 @@
+.tabs-component {
+  margin-bottom: govuk-spacing(4);
```
Comment (lsalivonaite-transform) [inline app/frontend/styles/components/_tabs.scss:2]: Do you think we could override this to 0? As it's creating unnecessarily huge gap between the navigation and the applications
Screenshot 2024-02-16 at 14 40 53

Comment (tomas-stefano) [reply app/frontend/styles/components/_tabs.scss:2]: Yes, I will do

Comment (tomas-stefano) [reply app/frontend/styles/components/_tabs.scss]: Done!

File: app/frontend/styles/components/_tabs.scss (hunk 1 above)
Comment (inulty-dfe) [reply app/frontend/styles/components/_tabs.scss:54]: I don't think this is a problem

**Review by lsalivonaite-transform — APPROVED** [review]
It seems to be all working, I have tested this in Safari, Chrome and Edge, didnt encounter any issues!
The only thing that seems to be odd, is the 'interviewing' status, I think it's supposed to be yellow if I am not mistaken, currently it's showing as purple.
Screenshot 2024-02-16 at 15 33 26
Screenshot 2024-02-16 at 15 33 42

File: config/locales/candidate_interface/application_choice_list.yml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+en:
+  candidate_interface:
+    application_tabs:
+      all: All applications
+      offers_received: Offers received
+      draft: Draft
+      unsuccessful: Unsuccessful
+      in_progress: In progress
+      withdraw: Withdraw
```
Comment (JR-G) [inline config/locales/candidate_interface/application_choice_list.yml:9]: Should this be `Withdrawn`?

Comment (tomas-stefano) [reply config/locales/candidate_interface/application_choice_list.yml]: Good spot. Fixed.

**Review by JR-G — APPROVED** [review]

File: app/frontend/styles/_application-item.scss [hunk 5]
```diff
@@ -41,7 +43,7 @@
   @include govuk-media-query($from: tablet) {
     flex-direction: row-reverse;
     margin-bottom: 0;
-    margin-left: govuk-spacing(2);
+    padding-left: 10px;
```
Comment (inulty-dfe) [inline app/frontend/styles/_application-item.scss:46]: This should be `govuk-spacing(2)`

File: app/services/candidate_interface/sort_application_choices.rb [hunk 6]
```diff
@@ -4,19 +4,28 @@ def self.call(application_choices:)
...
           SELECT application.*,
+          CASE
+            WHEN (status IN ('offer')) THEN 'offers_received'
+            WHEN (status IN ('unsubmitted', 'application_not_sent', 'cancelled')) THEN 'draft'
+            WHEN (status IN ('rejected', 'conditions_not_met')) THEN 'unsuccessful'
+            WHEN (status IN ('interviewing', 'inactive', 'awaiting_provider_decision')) THEN 'in_progress'
+            WHEN (status IN ('offer_withdrawn', 'withdrawn')) THEN 'withdrawn'
+            WHEN (status IN ('declined')) THEN 'declined'
+            ELSE ''
+          END AS application_choices_group,
```
Comment (inulty-dfe) [inline app/services/candidate_interface/sort_application_choices.rb:15]: `application_choices_group_name` might be better considering the other column is called `application_choices_group_number`

File: spec/system/candidate_interface/continuous_applications/course_selection/candidate_selecting_a_course_spec.rb [hunk 7]
```diff
@@ -158,7 +158,13 @@ def when_the_course_is_full
   end
 
   def when_i_visit_the_course_choices_page
-    visit candidate_interface_continuous_applications_choices_path
+    @application_choice = current_candidate.current_application.application_choices.first
+    when_i_visit_my_applications
+  end
+
+  def then_i_should_see_that_the_course_is_full
```
Comment (inulty-dfe) [inline spec/system/candidate_interface/continuous_applications/course_selection/candidate_selecting_a_course_spec.rb:165]: You've added a `should` here :grimacing:

Comment (tomas-stefano) [reply spec/system/candidate_interface/continuous_applications/course_selection/candidate_selecting_a_course_spec.rb:165]: I know, I was like, "should" I continue to follow the "should" until we remove it or not? 😢

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_sees_all_their_applications_spec.rb [hunk 8]
```diff
@@ -0,0 +1,137 @@
+require 'rails_helper'
```
Comment (inulty-dfe) [inline spec/system/candidate_interface/continuous_applications/submitting/candidate_sees_all_their_applications_spec.rb:1]: This file is in the submitting directory but no applications are being submitted. I think we should put this in a different directory

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_sees_all_their_applications_spec.rb [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    when_i_click_to_view_my_application
+    then_i_should_be_in_view_my_application_page
+    when_i_visit_my_applications
+    and_i_click_to_view_the_offers_tab
+    then_i_should_see_only_the_offers_applications
+    when_i_click_to_view_my_application
+    then_i_should_be_on_the_details_offer_page
+    when_i_visit_my_applications
+    and_i_click_to_view_the_unsuccessful_tab
+    then_i_should_see_only_the_unsuccessful_applications
```
Comment (inulty-dfe) [inline spec/system/candidate_interface/continuous_applications/submitting/candidate_sees_all_their_applications_spec.rb:34]: So many `should`s :1234:

File: app/frontend/styles/_application-item.scss (hunk 5 above)
Comment (tomas-stefano) [reply app/frontend/styles/_application-item.scss]: Done

File: app/services/candidate_interface/sort_application_choices.rb (hunk 6 above)
Comment (tomas-stefano) [reply app/services/candidate_interface/sort_application_choices.rb]: Good suggestion. Done

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_sees_all_their_applications_spec.rb (hunk 8 above)
Comment (tomas-stefano) [reply spec/system/candidate_interface/continuous_applications/submitting/candidate_sees_all_their_applications_spec.rb]: Good suggestion. I move to `spec/system/candidate_interface/continuous_applications/viewing-application-choices`.

**Review by inulty-dfe — APPROVED** [review]

**Pull request merged**